### ISSUE TITLE
PointNet implementation

### ISSUE THREAD

**Issue by sunshantong** [j96w/DenseFusion#34]
Hi, a _PointNet-based_ network that processes each point in the masked 3D point cloud to a geometric feature embedding, said in the paper. But in the code implementation only convolution and ReLU of point cloud _x_ in _network.py_. No spacial transform networks(STN) and maxpooling operation on point cloud. These are the two most important features of PointNet.

` Class PoseNetFeat (nn. Module):
     def __init__(self, num_points):
        super(PoseNetFeat, self).__init__()
        self.conv1 = torch.nn.Conv1d(3, 64, 1)
        self.conv2 = torch.nn.Conv1d(64, 128, 1)

        self.e_conv1 = torch.nn.Conv1d(32, 64, 1)
        self.e_conv2 = torch.nn.Conv1d(64, 128, 1)

        self.conv5 = torch.nn.Conv1d(256, 512, 1)
        self.conv6 = torch.nn.Conv1d(512, 1024, 1)

        self.ap1 = torch.nn.AvgPool1d(num_points)
        self.num_points = num_points
    def forward(self, x, emb):
        x = F.relu(self.conv1(x))
        emb = F.relu(self.e_conv1(emb))
        pointfeat_1 = torch.cat((x, emb), dim=1)

        x = F.relu(self.conv2(x))
        emb = F.relu(self.e_conv2(emb))
        pointfeat_2 = torch.cat((x, emb), dim=1)

        x = F.relu(self.conv5(pointfeat_2))
        x = F.relu(self.conv6(x))

        ap_x = self.ap1(x)

        ap_x = ap_x.view(-1, 1024, 1).repeat(1, 1, self.num_points)
        return torch.cat([pointfeat_1, pointfeat_2, ap_x], 1) #128 + 256 + 1024`
I am confused. Would you explain this to me?
Thanks in advance.

**Comment by j96w** [comment]
Hi,
(1) Since we are already uncovering the 6DoF pose (rotation and translation) of the target object, we don't need the STN to rotate the input pointcloud which might disturb the pose estimation process.
(2) Maxpooling will lead to poor training efficiency of the CNN-based encoder-decoder network since it's skipping too many pixel embeddings during the Max-pool step.

**Comment by sunshantong** [comment]
Thank you for your reply. And what about scale-awareness like PointSIFT or sampling&grouping like PoseNet++? Do you think it will be helpful for pointcloud processing? Looking forward to your suggestions.

**Comment by j96w** [comment]
Ya, PointNet++ might worth a try.

**Comment by sunshantong** [comment]
Thanks.

**Comment by trevor-taeyeop** [comment]
> Hi, a _PointNet-based_ network that processes each point in the masked 3D point cloud to a geometric feature embedding, said in the paper. But in the code implementation only convolution and ReLU of point cloud _x_ in _network.py_. No spacial transform networks(STN) and maxpooling operation on point cloud. These are the two most important features of PointNet.
> 
> ` Class PoseNetFeat (nn. Module):
> def **init**(self, num_points):
> super(PoseNetFeat, self).**init**()
> self.conv1 = torch.nn.Conv1d(3, 64, 1)
> self.conv2 = torch.nn.Conv1d(64, 128, 1)
> 
> ```
>     self.e_conv1 = torch.nn.Conv1d(32, 64, 1)
>     self.e_conv2 = torch.nn.Conv1d(64, 128, 1)
> 
>     self.conv5 = torch.nn.Conv1d(256, 512, 1)
>     self.conv6 = torch.nn.Conv1d(512, 1024, 1)
> 
>     self.ap1 = torch.nn.AvgPool1d(num_points)
>     self.num_points = num_points
> def forward(self, x, emb):
>     x = F.relu(self.conv1(x))
>     emb = F.relu(self.e_conv1(emb))
>     pointfeat_1 = torch.cat((x, emb), dim=1)
> 
>     x = F.relu(self.conv2(x))
>     emb = F.relu(self.e_conv2(emb))
>     pointfeat_2 = torch.cat((x, emb), dim=1)
> 
>     x = F.relu(self.conv5(pointfeat_2))
>     x = F.relu(self.conv6(x))
> 
>     ap_x = self.ap1(x)
> 
>     ap_x = ap_x.view(-1, 1024, 1).repeat(1, 1, self.num_points)
>     return torch.cat([pointfeat_1, pointfeat_2, ap_x], 1) #128 + 256 + 1024`
> ```
> 
> I am confused. Would you explain this to me?
> Thanks in advance.

I have some related question about this part.

1. I'm confused in this code which variable represents color embeddings and geometry embeddings??
 I understood pointfeat_1(pixel), pointfeat_2(color and geometry embeding_space), ap_x(global_feature). Is is right? 

Would you explain about this part?? 
Thanks.

**Comment by 704051419** [comment]
though STN and symmetry function like max pooling are the key idea of PointNet, cause they are the operation to deal with the  problem of pointcloud data , otherwise the whole architecture is actually not so "PointNet" ...
